wallet/src/repository/wallet/mod.rs

```diff
@@ -1668,14 +1668,22 @@ where
         let mut state = self.state.write()?;
 
         // Prevent chain data from being cleared if a sync or resync process is in progress
-        if !state.is_syncing() {
+        if !state.synchronization.is_resyncing() {
             state.clear_chain_data(&self.params.genesis_prev_hash);
 
             Ok(true)
         } else {
             Ok(false)
         }
     }
+
+    /// Run a predicate on the state of a wallet in a thread safe manner, thanks to a write lock.
+    pub fn apply_on_state<P, O>(&self, predicate: P) -> Result<O>
```

Comment:
This is a bad idea because this function can only succeed if the write lock was not held, and there are many functions in our code that take a write lock at the top of the function. This is why there are function names that start with `_`. I would prefer to be explicit about the lock, and therefore remove this function.

Reply:
I see that it's useful, so we could rename it to something that mentions a write lock. `apply_on_state_write_lock` for example

Reply:
I get your point about the write lock. This is sort of dangerous because it can lead to code that looks right at compile time but then causes a deadlock. However, every each of the functions that start with `_` have their public counterparts that acquire a lock in exactly this same way. I don't see why this would be worse than those.